src/Algebra/Morphism/Bundles.agda

```diff
@@ -0,0 +1,223 @@
+-- Morphisms between NearSemirings
+------------------------------------------------------------------------
+
+record NearSemiringHomomorphism (A : NearSemiring a ℓa) (B : NearSemiring b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = NearSemiring A
+    module B = NearSemiring B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+
+    isNearSemiringHomomorphism : IsNearSemiringHomomorphism A.rawNearSemiring B.rawNearSemiring ⟦_⟧
+
+  open IsNearSemiringHomomorphism isNearSemiringHomomorphism public
+
+  +-monoidHomomorphism : MonoidHomomorphism A.+-monoid B.+-monoid
+  +-monoidHomomorphism = record { isMonoidHomomorphism = +-isMonoidHomomorphism }
+
+  *-magmaHomomorphism : MagmaHomomorphism A.*-magma B.*-magma
+  *-magmaHomomorphism = record { isMagmaHomomorphism = *-isMagmaHomomorphism }
+
+------------------------------------------------------------------------
+-- Morphisms between Semirings
+------------------------------------------------------------------------
+
+record SemiringHomomorphism (A : Semiring a ℓa) (B : Semiring b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = Semiring A
+    module B = Semiring B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+
+    isSemiringHomomorphism : IsSemiringHomomorphism A.rawSemiring B.rawSemiring ⟦_⟧
+
+  open IsSemiringHomomorphism isSemiringHomomorphism public
+
+  nearSemiringHomomorphism : NearSemiringHomomorphism A.nearSemiring B.nearSemiring
+  nearSemiringHomomorphism = record { isNearSemiringHomomorphism = isNearSemiringHomomorphism }
+
+  open NearSemiringHomomorphism nearSemiringHomomorphism public
+    using (*-magmaHomomorphism; +-monoidHomomorphism)
+
+  *-monoidHomomorphism : MonoidHomomorphism A.*-monoid B.*-monoid
+  *-monoidHomomorphism = record { isMonoidHomomorphism = *-isMonoidHomomorphism }
+
+------------------------------------------------------------------------
+-- Morphisms between KleeneAlgebras
+------------------------------------------------------------------------
+
+record KleeneAlgebraHomomorphism (A : KleeneAlgebra a ℓa) (B : KleeneAlgebra b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
+  private
+    module A = KleeneAlgebra A
+    module B = KleeneAlgebra B
+
+  field
+    ⟦_⟧ : A.Carrier → B.Carrier
+    isKleeneAlgebraHomomorphism : IsKleeneAlgebraHomomorphism A.rawKleeneAlgebra B.rawKleeneAlgebra ⟦_⟧
+
+  open IsKleeneAlgebraHomomorphism isKleeneAlgebraHomomorphism public
+
+  semiringHomomorphism : SemiringHomomorphism A.semiring B.semiring
+  semiringHomomorphism = record { isSemiringHomomorphism = isSemiringHomomorphism }
+
+  open SemiringHomomorphism semiringHomomorphism public
+    hiding (*-isMagmaHomomorphism; *-isMonoidHomomorphism)
+
+------------------------------------------------------------------------
+-- Morphisms between RingWithoutOnes
+------------------------------------------------------------------------
+
+record RingWithoutOneHomomorphism (A : RingWithoutOne a ℓa) (B : RingWithoutOne b ℓb) : Set (a ⊔ b ⊔ ℓa ⊔ ℓb) where
```

Comment:
This should expose
* `nearSemiringHomomorphism`
* `+-magmaHomomorphism`
* `+-monoidHomomorphism`

Reply:
This turned out to be quite a bit more far-reaching than I first suspected... see incoming commits.